toolchain/sem_ir/file.cpp

```diff
@@ -308,214 +309,135 @@ static auto StringifyTypeExprImpl(const SemIR::File& outer_sem_ir,
       steps.push_back({.sem_ir = sem_ir, .inst_id = inst_id});
     };
 
-    auto inst = sem_ir.insts().Get(step.inst_id);
-    switch (inst.kind()) {
-      case ArrayType::Kind: {
-        auto array = inst.As<ArrayType>();
-        if (step.index == 0) {
-          out << "[";
-          steps.push_back(step.Next());
-          push_inst_id(sem_ir.types().GetInstId(array.element_type_id));
-        } else if (step.index == 1) {
-          out << "; " << sem_ir.GetArrayBoundValue(array.bound_id) << "]";
-        }
-        break;
-      }
-      case AssociatedEntityType::Kind: {
-        auto assoc = inst.As<AssociatedEntityType>();
-        if (step.index == 0) {
-          out << "<associated ";
-          steps.push_back(step.Next());
-          push_inst_id(sem_ir.types().GetInstId(assoc.entity_type_id));
-        } else {
+    TypedInstSwitch(sem_ir.insts().Get(step.inst_id))
+        .Case([&](ArrayType inst) {
+          if (step.index == 0) {
+            out << "[";
+            steps.push_back(step.Next());
```

Comment:
The lambda bodies end up indented three steps from the switch call here. I wonder if we can do better.

I've looked at some [other syntactic options](https://godbolt.org/z/a1YPnTn15) to try to reduce the indentation here:

- (This PR, LLVM `StringSwitch`-style): `Switch(arg).Case(a).Case(b)`: 3 step indentation to lambda body.
- Variadic call: `Switch(arg, a, b)`: 3 step indentation to lambda body. However, may make table dispatch easier to write.
- `Switch(arg) | a | b`: 1 step indentation to lambda body. `clang-format` gives this fairly dense formatting:

    ```c++
    Switch(arg) | [&](TypeA inst) {
      // ...
    } | [&](TypeB inst) {
      // ...
    } | [&](TypeC inst) {
      // ...
    };
    ```
    Most other non-keyword binary operators (`*`, `||`, `^`, `+`, `=`, `<<`, `>>`) give similar results.
- `Switch(arg) or a or b`: 1 step indentation to lambda body. Less densely formatted:

    ```c++
    Switch(arg) | [&](TypeA inst) {
      // ...
    }
    or [&](TypeB inst) {
      // ...
    }
    or [&](TypeC inst) {
      // ...
    };
    ```
- `Switch(arg), a, b`: 3 step indent to lambda body.

I think the non-keyword binary operator option gives the best formatting here -- it's reasonably dense, and there are no `}` lines that might suggest we're at the end of the statement. But are we prepared to use an operator for this?

Reply:
For the shorter lambda bodies in `GetExprCategory`, we get [a formatting pretty similar to with `.Case`](https://godbolt.org/z/o3hqWqoWz) if we use a binary operator.

Reply:
The variadic might be appealing due to the `()` wrapping the full body of the switch, similar to an actual switch statement. In general, I'm a bit cautious about the understandability of passing lambdas to operators, and `or` in particular I think might not carry the right semantic for readers. I also think at least _some_ indentation makes it easier to identify the start and end.

That said regarding indentation, comparing your two example code links, you seem to get significantly different indentation for the `|` operator. I wonder if this may not be working as well as hoped? The inconsistency may also make it less desirable.

One thing that makes me hesitant about these alternatives is how they handle the `Cases` equivalent, where it might cast but not use the result. Maybe the variadic approach would make writing a dispatch table there, which I suspect we might like to end up at even though I'm not writing it here.

Either way, I'm currently pausing on digging into the syntax until we can get agreement at least in _principle_. If Chandler has a strong preference towards a `switch` statement, I think that would nix work.